File: python/mirage/commands/builtin/ram/gzip.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.accessor.ram import RAMAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.gzip import extract_level
from mirage.commands.builtin.generic.gzip import gzip as generic_gzip
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.ram.glob import resolve_glob
from mirage.core.ram.read import read_bytes
from mirage.core.ram.unlink import unlink
from mirage.core.ram.write import write_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("gzip", resource="ram", spec=SPECS["gzip"], write=True)
async def gzip(
    accessor: RAMAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    d: bool = False,
    k: bool = False,
    f: bool = False,
    c: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    level = extract_level(_extra)
    if paths:
        paths = await resolve_glob(accessor, paths, index)
    return await generic_gzip(paths,
                              read_bytes=read_bytes,
                              write_bytes=write_bytes,
                              unlink=unlink,
                              accessor=accessor,
                              stdin=stdin,
                              decompress=d,
                              keep=k,
                              force=f,
                              to_stdout=c,
                              level=level)

File: typescript/packages/node/src/commands/builtin/redis/gzip.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { ResourceName, command, specOf, gzipGeneric } from '@struktoai/mirage-core'
import type { RedisAccessor } from '../../../accessor/redis.ts'
import { stream as redisStream } from '../../../core/redis/stream.ts'
import { writeBytes as redisWrite } from '../../../core/redis/write.ts'
import { unlink as redisUnlink } from '../../../core/redis/unlink.ts'

export const REDIS_GZIP = command({
  name: 'gzip',
  resource: ResourceName.REDIS,
  spec: specOf('gzip'),
  fn: (accessor: RedisAccessor, paths, _texts, opts) =>
    gzipGeneric(
      paths,
      opts,
      (p) => redisStream(accessor, p),
      (p, d) => redisWrite(accessor, p, d),
      (p) => redisUnlink(accessor, p),
    ),
  write: true,
})
